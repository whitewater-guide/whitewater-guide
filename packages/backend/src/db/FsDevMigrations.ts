// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { FsMigrations } from 'knex/lib/migrate/sources/fs-migrations';

// When restoring production dump on development, migration names will have '.js' extension
// Therefore, development migrations with '.ts' will conflict them
// this is the workaround
// https://github.com/knex/knex/blob/master/lib/migrate/sources/fs-migrations.js
export class FsDevMigrations extends FsMigrations {
  // eslint-disable-next-line @typescript-eslint/no-useless-constructor
  constructor(
    migrationDirectories: string | string[],
    sortDirsSeparately?: boolean,
    loadExtensions?: string[],
  ) {
    super(migrationDirectories, sortDirsSeparately, loadExtensions);
  }

  getMigrationName(migration: any): string {
    return migration.file.replace('.ts', '.js');
  }
}
