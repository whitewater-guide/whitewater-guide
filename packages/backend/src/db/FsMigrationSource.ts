import fs from 'node:fs/promises';
import path from 'node:path';

import type { Knex } from 'knex';

// https://knexjs.org/guide/migrations.html#custom-migration-sources
export default class FsMigrationSource implements Knex.MigrationSource<string> {
  // eslint-disable-next-line node/no-process-env
  private readonly ext = process.env.NODE_ENV === 'production' ? 'js' : 'ts';
  private readonly directory: string;

  constructor(directory: string) {
    this.directory = directory;
  }
  // Must return a Promise containing a list of migrations.
  // Migrations can be whatever you want,
  // they will be passed as arguments to getMigrationName
  // and getMigration
  //
  // In our case migrations are filenames without extension
  async getMigrations(): Promise<string[]> {
    // In this example we are just returning migration names
    const dirContent = await fs.readdir(this.directory);
    // assume that migrations have extensions (== no directories with '.' symbol)
    return dirContent
      .filter((f) => f.includes('.'))
      .map((f) => f.split('.')[0])
      .sort();
  }

  // When restoring production dump on development, migration names will have '.js' extension
  getMigrationName(migration: string): string {
    return migration + '.js';
  }

  async getMigration(migration: string): Promise<Knex.Migration> {
    const moduleName = path.resolve(
      process.cwd(),
      this.directory,
      `${migration}.${this.ext}`,
    );
    return import(moduleName);
  }
}
