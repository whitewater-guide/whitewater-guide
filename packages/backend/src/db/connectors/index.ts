export * from './createConnectors';
export * from './OffsetConnector';
export * from './types';
