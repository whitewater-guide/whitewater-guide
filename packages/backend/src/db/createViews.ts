import { readFileSync } from 'fs';
import Knex from 'knex';
import path from 'path';

export async function createView(db: Knex, view: string, version = 1) {
  for (let i = version; i >= 1; i--) {
    const file = path.resolve(
      __dirname,
      '../migrations',
      `${i.toString(10).padStart(3, '0')}/${view}_view.sql`,
    );
    let sql: string;
    try {
      sql = readFileSync(file).toString();
    } catch (e) {
      continue;
    }
    await db.schema.raw(sql);
    return;
  }
  throw new Error(
    `No sql files for view '${view}' of version ${version} or lower found`,
  );
}

// Views will be created in reverse order
export async function createViews(
  db: Knex,
  version: number,
  ...views: string[]
) {
  const inverted = [...views].reverse();
  for (const view of inverted) {
    await createView(db, view, version);
  }
}
