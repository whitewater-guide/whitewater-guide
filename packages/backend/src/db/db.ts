/* eslint-disable node/no-process-env */
import Knex from 'knex';

import knexConfig from './knexfile';

const env = process.env.NODE_ENV || 'development';
const config: Knex.Config = knexConfig[env];

export const knex = Knex(config);

let testTransaction: Knex.Transaction | null = null;

export function db(getKnexInstance = false): Knex<any, any> {
  if (env === 'test' && !getKnexInstance) {
    if (testTransaction) {
      return testTransaction;
    }
    throw new Error('No saved test transaction found');
  }
  return knex;
}

// See here https://github.com/tgriesser/knex/issues/2076
// And here https://gist.github.com/odigity/7f37077de74964051d45c4ca80ec3250
export const holdTransaction = () => {
  if (env !== 'test') {
    throw new Error('holdTransaction is only available in test mode');
  }
  const tmp: any = {};
  const p = new Promise((resolve) => {
    tmp.resolve = resolve;
  });
  knex
    .transaction((tx) => {
      testTransaction = tx;
      tmp.resolve();
    })
    .catch(() => {
      // ignore
    });
  return p;
};

export const rollbackTransaction = async () => {
  if (env !== 'test') {
    throw new Error('rollbackTransaction is only available in test mode');
  }
  if (testTransaction) {
    await testTransaction.rollback();
    testTransaction = null;
  }
};
