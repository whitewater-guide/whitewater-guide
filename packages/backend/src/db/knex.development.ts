import knex from 'knex';

const development: knex.Config = {
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST!,
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
    keepAlive: true,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './src/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './src/seeds/development',
  },
  debug: process.env.DATABASE_DEBUG === 'true',
} as any;

module.exports = development;
