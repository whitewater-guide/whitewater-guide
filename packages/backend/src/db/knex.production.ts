import knex from 'knex';

const development: knex.Config = {
  client: 'pg',
  connection: {
    host: process.env.POSTGRES_HOST!,
    port: 5432,
    database: process.env.POSTGRES_DB!,
    user: 'postgres',
    password: process.env.POSTGRES_PASSWORD!,
    keepAlive: true,
  },
  migrations: {
    loadExtensions: ['.js'],
    tableName: 'migrations',
    directory: './dist/migrations',
  },
  seeds: {
    loadExtensions: ['.js'],
    directory: './dist/seeds/production',
  },
  pool: {
    min: 0,
  },
  // debug: true,
} as any;

module.exports = development;
