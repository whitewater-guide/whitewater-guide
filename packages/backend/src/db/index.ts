import { db, holdTransaction, knex, rollbackTransaction } from './db';

export * from './createViews';
export * from './dropViews';
export * from './rawUpsert';
export * from './runSqlFile';
export * from './types';
export * from './updatedAtTrigger';
export * from './waitForDb';
export { holdTransaction, rollbackTransaction, knex };

export default db;
