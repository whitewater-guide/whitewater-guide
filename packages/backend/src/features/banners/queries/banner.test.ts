import { anonContext, fakeContext, runQuery } from '@test';
import { ApolloErrorCodes } from '@whitewater-guide/commons';

import { holdTransaction, rollbackTransaction } from '~/db';
import { ADMIN, EDITOR_GA_EC, TEST_USER } from '~/seeds/test/01_users';
import { GROUP_ALL } from '~/seeds/test/03_groups';
import { REGION_GALICIA } from '~/seeds/test/04_regions';
import {
  ALL_SECTION_ROW_BANNER,
  GALICIA_REGION_DESCR_BANNER2,
  GALICIA_SECTION_ROW_BANNER,
} from '~/seeds/test/14_banners';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const query = `
  query bannerDetails($id: ID){
    banner(id: $id) {
      id
      name
      slug
      priority
      enabled
      placement
      source {
        kind
        url(width: 1000)
      }
      link
      extras
      regions {
        nodes {
          id
          name
        }
        count
      }
      groups {
        nodes {
          id
          name
        }
        count
      }
    }
  }
`;

describe('permissions', () => {
  it('anon shall fail', async () => {
    const result = await runQuery(
      query,
      { id: GALICIA_SECTION_ROW_BANNER },
      anonContext(),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.UNAUTHENTICATED);
  });

  it('user shall not pass', async () => {
    const result = await runQuery(
      query,
      { id: GALICIA_SECTION_ROW_BANNER },
      fakeContext(TEST_USER),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });

  it('editor shall not pass', async () => {
    const result = await runQuery(
      query,
      { id: GALICIA_SECTION_ROW_BANNER },
      fakeContext(EDITOR_GA_EC),
    );
    expect(result).toHaveGraphqlError(ApolloErrorCodes.FORBIDDEN);
  });
});

describe('data', () => {
  it('should return banner', async () => {
    const result = await runQuery(
      query,
      { id: GALICIA_SECTION_ROW_BANNER },
      fakeContext(ADMIN),
    );
    expect(result).not.toHaveGraphqlError();
    expect(result.data!.banner).toMatchSnapshot();
  });

  it('should return null when id not specified', async () => {
    const result = await runQuery(query, {}, fakeContext(ADMIN));
    expect(result).not.toHaveGraphqlError();
    expect(result.data!.banner).toBeNull();
  });

  it('should return thumb', async () => {
    const result = await runQuery(
      query,
      { id: GALICIA_REGION_DESCR_BANNER2 },
      fakeContext(ADMIN),
    );
    expect(result.errors).toBeUndefined();
    expect(result.data!.banner.source.url).toBe(
      'imgproxy://w:1024/banners/banner_4.jpg',
    );
  });
});

describe('connections', () => {
  describe('regions', () => {
    it('should return nodes', async () => {
      const q = `
      query bannerDetails($id: ID){
        banner(id: $id) {
          id
          name
          slug
          regions {
            count
            nodes {
              id
              name
              __typename
            }
            __typename
          }
          __typename
        }
      }
    `;
      const result = await runQuery(
        q,
        { id: GALICIA_SECTION_ROW_BANNER },
        fakeContext(ADMIN),
      );
      expect(result).not.toHaveGraphqlError();
      const banner = result.data!.banner;
      expect(banner).toMatchObject({
        id: GALICIA_SECTION_ROW_BANNER,
        name: 'galicia section row banner',
        regions: {
          count: 1,
          nodes: [{ id: REGION_GALICIA, name: 'Galicia' }],
        },
      });
    });
  });

  describe('groups', () => {
    it('should return nodes', async () => {
      const q = `
      query bannerDetails($id: ID){
        banner(id: $id) {
          id
          name
          slug
          groups {
            count
            nodes {
              id
              name
              __typename
            }
            __typename
          }
          __typename
        }
      }
    `;
      const result = await runQuery(
        q,
        { id: ALL_SECTION_ROW_BANNER },
        fakeContext(ADMIN),
      );
      expect(result).not.toHaveGraphqlError();
      const banner = result.data!.banner;
      expect(banner).toMatchObject({
        id: ALL_SECTION_ROW_BANNER,
        name: 'all section row banner',
        groups: {
          count: 1,
          nodes: [{ id: GROUP_ALL, name: 'All regions' }],
        },
      });
    });
  });
});
