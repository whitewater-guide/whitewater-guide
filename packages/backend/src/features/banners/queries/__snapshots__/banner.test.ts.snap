// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data should return banner 1`] = `
Object {
  "enabled": true,
  "extras": null,
  "groups": Object {
    "count": 0,
    "nodes": Array [],
  },
  "id": "577ef5e8-c66d-11e8-a355-529269fb1459",
  "link": "http://go.to/galicia_section_row_banner",
  "name": "galicia section row banner",
  "placement": "mobileSectionRow",
  "priority": 1,
  "regions": Object {
    "count": 1,
    "nodes": Array [
      Object {
        "id": "bd3e10b6-7624-11e7-b5a5-be2e44b06b34",
        "name": "Galicia",
      },
    ],
  },
  "slug": "galicia_section_row_banner",
  "source": Object {
    "kind": "WebView",
    "url": "http://whitewater.guide/galicia_section_row_banner",
  },
}
`;
